src/app/core/internal/impl/ComponentInfo.ts

```diff
@@ -0,0 +1,112 @@
+import { Schema } from "core/schema";
+
+export interface PortConfig {
+    counts: number[];
+};
+
+// Describes legal component configurations.
+// NOTE: The configuration wherein each port group has zero ports is IMPLICITLY LEGAL.
+export interface ComponentInfo {
+    readonly kind: string;
+    readonly portGroups: string[];
+    readonly defaultPortConfig: PortConfig;
+
+    isValidPortConfig(p: PortConfig): boolean;
+    makePortsForConfig(id: Schema.GUID, p: PortConfig): Schema.Port[] | undefined;
+
+    // i.e. Prevents fan-in on digital ports, could prevent illegal self-connectivity
+    isValidPortConnectivity(wires: Map<Schema.Port, Schema.Port[]>): boolean;
+
+    // Also i.e. thumbnail, display name, description, etc.
+}
+
+export interface ComponentInfoProvider {
+    get(kind: string): ComponentInfo | undefined;
+}
+
+// TODO: seems kind of heavy
+export function IsValidPortList(info: ComponentInfo, ports: Schema.Port[]): boolean {
+    let countMap = new Map(info.portGroups.map(g => [g, Number(0)]));
+    for (const p of ports) {
+        let count = countMap.get(p.group);
+        if (count === undefined)
+            return false;
+        count++;
+    }
+    let counts: number[] = info.portGroups.map(g => countMap.get(g) ?? -1);
+    return info.isValidPortConfig({ counts: counts });
```

Comment:
Mmmmm I don't like that `counts` needs to be an array that's 1:1 with the groups defined in `ComponentInfo`, can we not just pass this map of group types to counts in `info.isValidPortConfig`? see below